### ISSUE TITLE
LINQ IndexWhere extension method

### ISSUE THREAD

**Issue by masonwheeler** [dotnet/corefx#39689]
I recently found myself writing up yet another ad-hoc `IndexWhere` extension method for `IEnumerable<T>` in yet another project.  Every time the need for it comes up, I'm a bit surprised that it's never been put in `System.Linq.Enumerable`, given how common an operation it turns out to be.

We should fix that.  Let's add a standard LINQ operator that works exactly like `Where`, except that instead of returning `IEnumerable<T>`, it returns `IEnumerable<int>`, with the indices of all the values that match the predicate.

This can currently be emulated with existing standard LINQ operators, but the best solution I've found takes three: `input.Select((x, i) => new KeyValuePair<T, int>(x, i)).Where(kvp => predicateOperation(kvp.Key)).Select(kvp => kvp.Value)`, which is ugly and inefficient.  Let's do it in one instead.

Reference implementation, with error checking elided for clarity:

```
        public static IEnumerable<int> IndexWhere<T>(this IEnumerable<T> value, Func<T, bool> predicate)
        {
            int i = 0;
            foreach (var item in value)
            {
                if (predicate(item))
                {
                    yield return i;
                }
                ++i;
            }
        }
```

**Comment by EgorBo** [comment]
Shorter emulation:
```csharp
int? index = input.Select((x, i) => new {x, i}) 
                  .FirstOrDefault(i => predicateOperation(i.x))?.i;
```

**Comment by Gnbrkm41** [comment]
doesn't `Where(predicate).Select((x, i) => i)` just work?

**Comment by masonwheeler** [comment]
@Gnbrkm41 No, that will give a sequence `0, 1, 2, 3, ...` because `Select` only sees the items that passed the predicate successfully.

**Comment by Gnbrkm41** [comment]
Ah, true.

**Comment by TheFanatr** [comment]
While I agree that a method with this functionality could be nice, if there are enough common use cases, I feel the name `IndexWhere` is unsuitable as it implies that the functionality is "get the single index where this predicate is true". For this operation `IndexesWhere` may more easily obviate the true functionality, or even just `Index` as a verb because the functionality of the method is to "index elements in the list". Not sure if that last one really works like that though but I was mainly suggesting it because I dislike non-verb method names.

**Comment by Gnbrkm41** [comment]
How about `FindAll`?

**Comment by scalablecory** [comment]
Consider adding this to System.Interactive in https://github.com/dotnet/reactive, which has a lot of great utility extensions like this.

> Every time the need for it comes up, I'm a bit surprised that it's never been put in System.Linq.Enumerable, given how common an operation it turns out to be.

It's interesting for me to see how people use something like LINQ in such a diverse manner. I consider myself a heavy LINQ user and this is exceedingly rare for me.
